src/pkgcheck/checks/codingstyle.py

```diff
@@ -903,3 +903,118 @@ def feed(self, pkg):
                         yield RedundantDodir(
                             cmd.group('cmd'), line=dodir.group('call'),
                             lineno=lineno - 1, pkg=pkg)
+
+
+class UnquotedVariable(results.AliasResult, results.Warning):
+    """Variable is used unquoted in a context where it should be quoted.
+
+    Variables like D, FILESDIR, etc may not be safe to use unquoted in some
+    contexts.
+    """
+
+    _name = 'UnquotedVariable'
+
+    def __init__(self, variable, lines, **kwargs):
+        super().__init__(**kwargs)
+        self.variable = variable
+        self.lines = tuple(lines)
+
+    @property
+    def desc(self):
+        s = pluralism(self.lines)
+        lines = ', '.join(map(str, self.lines))
+        return f'unquoted variable {self.variable} on line{s}: {lines} '
+
+
+
+class EbuildUnquotedVariable(UnquotedVariable, results.VersionResult):
+    __doc__ = UnquotedVariable.__doc__
+
+
+class EclassUnquotedVariable(UnquotedVariable, results.EclassResult):
+    __doc__ = UnquotedVariable.__doc__
+
+    @property
+    def desc(self):
+        return f'{self.eclass}: {super().desc}'
+
+
+class _UnquotedVariablesCheck(Check):
+    """Scan files for variables that should be quoted like D, FILESDIR, etc."""
+
+    message_commands = (
+        "die", "echo", "eerror", "einfo", "elog", "eqawarn", "ewarn"
+    )
+    var_names = (
+        "D", "DISTDIR", "FILESDIR", "S", "T", "ROOT", "BROOT", "WORKDIR", "ED",
+        "EPREFIX", "EROOT", "SYSROOT", "ESYSROOT", "TMPDIR", "HOME",
+        # variables for multibuild.eclass
+        "BUILD_DIR",
+    )
```

Comment:
This is somewhat longer tuple, which we might need to increase it's size in future. So I think for faster lookup inside, let's use here `= frozenset({ ... })`